logging/nxscope/nxscope.c

```diff
@@ -0,0 +1,1138 @@
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <assert.h>
+#include <debug.h>
+#include <errno.h>
+#include <string.h>
+#include <stdlib.h>
+
+#include <logging/nxscope/nxscope.h>
+
+#include "nxscope_internals.h"
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: nxscope_frame_send
+ *
+ * NOTE: This function assumes that we have exclusive access to the nxscope
+ *       instance
+ *
+ ****************************************************************************/
+
+static int nxscope_frame_send(FAR struct nxscope_s *s, uint8_t id,
+                              FAR uint8_t *data, size_t dlen)
+{
+  size_t tx_i = 0;
+  int    ret  = OK;
+
+  DEBUGASSERT(s);
+  DEBUGASSERT(data);
+
+  /* Offset for hdr */
+
+  tx_i = s->proto_cmd->hdrlen;
+
+  /* Copy data */
+
+  memcpy(&s->txbuf[tx_i], data, dlen);
+  tx_i += dlen;
+
+  /* Finalize a new frame */
+
+  ret = PROTO_FRAME_FINAL(s, s->proto_cmd, id, s->txbuf, &tx_i);
+  if (ret < 0)
+    {
+      _err("ERROR: PROTO_FRAME_FINAL failed %d\n", ret);
+      goto errout;
+    }
+
+  /* Send frame */
+
+  ret = INTF_SEND(s, s->intf_cmd, s->txbuf, tx_i);
+  if (ret < 0)
+    {
+      _err("ERROR: INTF_SEND failed %d\n", ret);
+    }
+
+errout:
+  return ret;
+}
+
+/****************************************************************************
+ * Name: nxscope_cmninfo_send
+ *
+ * NOTE: This function assumes that we have exclusive access to the nxscope
+ *       instance
+ *
+ ****************************************************************************/
+
+static int nxscope_cmninfo_send(FAR struct nxscope_s *s)
+{
+  DEBUGASSERT(s);
+
+  return nxscope_frame_send(s,
+                            NXSCOPE_HDRID_CMNINFO,
+                            (FAR uint8_t *)&s->cmninfo,
+                            sizeof(struct nxscope_info_cmn_s));
+}
+
+/****************************************************************************
+ * Name: nxscope_chinfo_send
+ *
+ * NOTE: This function assumes that we have exclusive access to the nxscope
+ *       instance
+ *
+ ****************************************************************************/
+
+static int nxscope_chinfo_send(FAR struct nxscope_s *s, uint8_t ch)
+{
+  FAR uint8_t *data    = NULL;
+  size_t       namelen = 0;
+  size_t       txlen   = 0;
+  size_t       tmp     = 0;
+  int          ret     = OK;
+
+  DEBUGASSERT(s);
+
+  /* Allocate data for channel info */
+
+  namelen = strnlen(s->chinfo[ch].name, CHAN_NAMELEN_MAX) + 1;
+  tmp = sizeof(struct nxscope_chinfo_s) - sizeof(char *);
+  txlen = tmp  + namelen;
+
+  data = zalloc(txlen);
```

Comment:
Is that dynamic allocation really needed here? I don't know how often this is called, but it seems it will be called multiple times during system livetime. This will trash system memory on MCUs without MMU (memory fragmentation). It would be better to allocate `CHAN_NAMELEN_MAX + 1 + tmp` on stack since limits are well known during compile time.

Reply:
Good point. This dynamic allocation is a remnant of the previous version, where information about channels was sent in one large frame. But I decided to separate channels data and send them in separate frames. I'll fix it.

Reply:
I changed data to static buffer